chore: Add change log generation

richard-bogdan-bitrise opened the pull request:
Added internal feature of generating change log entries.

APM-2391

richard-bogdan-bitrise commented on build.gradle:
```diff
@@ -1,13 +1,21 @@
+        google()
+    }
+    dependencies {
+        classpath files('./buildSrc/build/classes/java/main')
```
This is needed, otherwise build.gradle won't be able to recognise the class.

richard-bogdan-bitrise commented on buildSrc/build.gradle:
```diff
@@ -0,0 +1,32 @@
+        jcenter()
+    }
+    dependencies {
+        classpath('org.eclipse.jgit:org.eclipse.jgit:5.10.0.202012080955-r')
```
Dependency responsible for doing Git actions.

daniellevass-bitrise replied:
such a peculiar naming strategy on their behalf! 🙃

richard-bogdan-bitrise replied:
> Looking great so far - i've left a few nits and comments around.
> 
> I think most importantly we need some better documentation - i cannot seem to run the tests on my machine. I've tried Android Studio / IntelliJ / updating android studio? Am i missing something? How are you running the tests?
> 
> Additionally, I ran the task not expecting the changelog to change and found it updated it to this:
> 
> <img alt="Screenshot 2021-02-24 at 11 40 20" width="1204" src="https://user-images.githubusercontent.com/71286749/108995982-c313ee00-7695-11eb-962d-2051b739397c.png">
> 
> is this expected? I thought [this line](https://github.com/bitrise-steplib/bitrise-add-trace-sdk-android/pull/16/files#diff-593183f3e1a6d6b5f5cf6be2d56c136ec1566cc6307a1a9d58066495e816bafeR112) would be reached and no changes found meant cancelling the task? We also don't have a test for this scenario so this might be worth adding :)



> Looking great so far - i've left a few nits and comments around.
> 
> I think most importantly we need some better documentation - i cannot seem to run the tests on my machine. I've tried Android Studio / IntelliJ / updating android studio? Am i missing something? How are you running the tests?
> 
> Additionally, I ran the task not expecting the changelog to change and found it updated it to this:
> 
> <img alt="Screenshot 2021-02-24 at 11 40 20" width="1204" src="https://user-images.githubusercontent.com/71286749/108995982-c313ee00-7695-11eb-962d-2051b739397c.png">
> 
> is this expected? I thought [this line](https://github.com/bitrise-steplib/bitrise-add-trace-sdk-android/pull/16/files#diff-593183f3e1a6d6b5f5cf6be2d56c136ec1566cc6307a1a9d58066495e816bafeR112) would be reached and no changes found meant cancelling the task? We also don't have a test for this scenario so this might be worth adding :)

Discussed on Slack and in the call the testing issue, sorted it out, updated documentation, thanks for mentioning this. 👍 

Yes, it is intended to happen to have maintenance release. It is in the case, when there were commits, but none of the are in the allowed types (fix, feat, fix!, feat!). Not sure if we will ever do a release when we have only for example test, chore  commits, but in my opinion it looks better than a missing entry. Of course I am open to discussion, let me know your thoughts.
I added a test case for this, please find it here:
[https://github.com/bitrise-steplib/bitrise-add-trace-sdk-android/blob/APM-2391/buildSrc/src/test/java/io/bitrise/trace/internal/UpdateChangeLogTaskTest.java#L385](https://github.com/bitrise-steplib/bitrise-add-trace-sdk-android/blob/APM-2391/buildSrc/src/test/java/io/bitrise/trace/internal/UpdateChangeLogTaskTest.java#L385)

richard-bogdan-bitrise replied:
Not sure why GH added this here, because I pushed on quote reply on the comment below 😕

richard-bogdan-bitrise commented on buildSrc/src/test/java/io/bitrise/trace/internal/UpdateChangeLogTaskTest.java:
```diff
@@ -0,0 +1,402 @@
+    }
+    // endRegion
+
+    // region getting commits
```
These tests in this region create a repo in a temporary folder with some dummy commits, this way it will be reproducible.

daniellevass-bitrise commented on buildSrc/src/main/java/io/bitrise/trace/internal/UpdateChangeLogTask.java:
```diff
@@ -0,0 +1,482 @@
+         *
+         * @return the String value of the current date.
+         */
+        private String getCurrentDate() {
```
not sure if we should have a consensus on the format of the date - the changelog for the trace-sdk looks like https://github.com/bitrise-io/trace-android-sdk/blob/main/CHANGELOG.md e.g. - 11th Feb 2021 - perhaps sticking with mm-dd-yyyy (and we'd need to update trace-sdk Feb to be 02)?

richard-bogdan-bitrise replied:
You are right, we haven't discussed this yet, happy to have a chat about it. I used the [ISO](https://en.wikipedia.org/wiki/ISO_8601) format, because that is the international standard format, but we should definitely keep this consistent across different products (even among teams). I think we should start a conversation about this with the team, but I am interested in your opinion, please let me know. 🙂

daniellevass-bitrise replied:
i agree with the international ISO8601 format too - i'm just not sure I like the dots 😅 Would you be adverse to hyphens e.g. 2021-02-24 instead of 2021.02.24?

richard-bogdan-bitrise replied:
Sure, updated it, nice catch!

daniellevass-bitrise reviewed (changes requested):
Looking great so far - i've left a few nits and comments around.

I think most importantly we need some better documentation - i cannot seem to run the tests on my machine. I've tried Android Studio / IntelliJ / updating android studio? Am i missing something? How are you running the tests?

Additionally, I ran the task not expecting the changelog to change and found it updated it to this:

<img width="1204" alt="Screenshot 2021-02-24 at 11 40 20" src="https://user-images.githubusercontent.com/71286749/108995982-c313ee00-7695-11eb-962d-2051b739397c.png">

is this expected? I thought [this line](https://github.com/bitrise-steplib/bitrise-add-trace-sdk-android/pull/16/files#diff-593183f3e1a6d6b5f5cf6be2d56c136ec1566cc6307a1a9d58066495e816bafeR112) would be reached and no changes found meant cancelling the task? We also don't have a test for this scenario so this might be worth adding :)

daniellevass-bitrise commented:
I just checked the acceptance criteria for this task, we're missing:
>Workflow step should run the tool to update the changelog.

I guess that might be kind of hard to run, and i'm happy for this to be a manual step - but can we ensure the tests are run regularly? either whenever a PR is raised, or a commit is merged into main?

richard-bogdan-bitrise commented on buildSrc/build.gradle:
```diff
@@ -0,0 +1,32 @@
+    testImplementation "org.hamcrest:hamcrest:2.2"
+}
+
+
```
Sure, thanks

richard-bogdan-bitrise commented on buildSrc/src/main/java/io/bitrise/trace/internal/UpdateChangeLogTask.java:
```diff
@@ -0,0 +1,482 @@
+            final String[] versionNumbers = version.split("\\.");
+            if (typeSet.stream().anyMatch(majorCommitTypes::contains)) {
+                logger.debug("New version will have major version increase!");
+                versionNumbers[0] = String.valueOf(Integer.parseInt(versionNumbers[0]) + 1);
```
Good point, updated!

richard-bogdan-bitrise commented on buildSrc/src/main/java/io/bitrise/trace/internal/UpdateChangeLogTask.java:
```diff
@@ -0,0 +1,482 @@
+import javax.inject.Inject;
+
+/**
+ * Internal task for updating the CHANGELOG.md of this repository. Run this task before release.
```
Added more details here. Will also add some details to the internal confluence docs.

daniellevass-bitrise replied:
I've noticed you updated the release process confluence doc - however it still doesn't explicitly tell people how e.g. `./gradlew UpdateChangeLogTask` doesn't run the task - what command do i need to type so i don't have to open android studio/intellij to run the task?

richard-bogdan-bitrise commented on buildSrc/build.gradle:
```diff
@@ -0,0 +1,32 @@
+buildscript {
```
buildSrc is a special directory, which is built always before any project task is executed. This is required to ensure that the UpdateChangeLogTask is present and can be referenced in the root build.gradle.

daniellevass-bitrise replied:
thanks for sharing the knowledge - i didn't know this and it was good to learn!

richard-bogdan-bitrise commented:
> I just checked the acceptance criteria for this task, we're missing:
> 
> > Workflow step should run the tool to update the changelog.
> 
> I guess that might be kind of hard to run, and i'm happy for this to be a manual step - but can we ensure the tests are run regularly? either whenever a PR is raised, or a commit is merged into main?

It would be relatively easy to add this to the workflow (one task call), and commit and push the changes automatically (2 lines of script, or as we use JGIT, we can even do it in the task), but not sure if we should do this. I think it would be better to have the manual control over this, at least in the beginning to see how it works. Which option do you prefer?

richard-bogdan-bitrise commented:
> Looking great so far - i've left a few nits and comments around.
> 
> I think most importantly we need some better documentation - i cannot seem to run the tests on my machine. I've tried Android Studio / IntelliJ / updating android studio? Am i missing something? How are you running the tests?
> 
> Additionally, I ran the task not expecting the changelog to change and found it updated it to this:
> 
> <img alt="Screenshot 2021-02-24 at 11 40 20" width="1204" src="https://user-images.githubusercontent.com/71286749/108995982-c313ee00-7695-11eb-962d-2051b739397c.png">
> 
> is this expected? I thought [this line](https://github.com/bitrise-steplib/bitrise-add-trace-sdk-android/pull/16/files#diff-593183f3e1a6d6b5f5cf6be2d56c136ec1566cc6307a1a9d58066495e816bafeR112) would be reached and no changes found meant cancelling the task? We also don't have a test for this scenario so this might be worth adding :)

Discussed on Slack and in the call the testing issue, sorted it out, updated documentation, thanks for mentioning this. 👍 

Yes, it is intended to happen to have maintenance release. It is in the case, when there were commits, but none of the are in the allowed types (fix, feat, fix!, feat!). Not sure if we will ever do a release when we have only for example test, chore  commits, but in my opinion it looks better than a missing entry. Of course I am open to discussion, let me know your thoughts.
I added a test case for this, please find it here:
[https://github.com/bitrise-steplib/bitrise-add-trace-sdk-android/blob/APM-2391/buildSrc/src/test/java/io/bitrise/trace/internal/UpdateChangeLogTaskTest.java#L385](https://github.com/bitrise-steplib/bitrise-add-trace-sdk-android/blob/APM-2391/buildSrc/src/test/java/io/bitrise/trace/internal/UpdateChangeLogTaskTest.java#L385)

richard-bogdan-bitrise commented on buildSrc/src/test/java/io/bitrise/trace/internal/UpdateChangeLogTaskTest.java:
```diff
@@ -0,0 +1,402 @@
+
+    private static final String dummyCommitTemplate = "%s: %s\n\n%s\n\n%s";
+
+    private static final String dummyCommitType1 = "fix";
```
I have to keep most of the strings, as I have to test that the created changelog entry matches the desired (please check dummyCommitType1 usages for example). I can create an object for it to save some lines, but I think then I would have to call toString() in the place of usages, so it is not a clear win.

daniellevass-bitrise replied:
hmm. I understand. I think for conciseness it would be better to do in the future, but i'm okay with how this is for now.

daniellevass-bitrise commented:
> > I just checked the acceptance criteria for this task, we're missing:
> > > Workflow step should run the tool to update the changelog.
> > 
> > 
> > I guess that might be kind of hard to run, and i'm happy for this to be a manual step - but can we ensure the tests are run regularly? either whenever a PR is raised, or a commit is merged into main?
> 
> It would be relatively easy to add this to the workflow (one task call), and commit and push the changes automatically (2 lines of script, or as we use JGIT, we can even do it in the task), but not sure if we should do this. I think it would be better to have the manual control over this, at least in the beginning to see how it works. Which option do you prefer?

I agree with the manual approach in regards to running the task. 
Could we update the workflow to run the tests?

daniellevass-bitrise reviewed (approved):
Looking really good.

Left a comment about being more explicit with how to run the task, perhaps using hyphens instead of dots for the date/time, and ensuring the tests for this run regularly - but none of those are blockers for this.

richard-bogdan-bitrise commented:
As discussed on the meeting, I will add a fix tomorrow to avoid case sensitivity issues for allowed commit types.

richard-bogdan-bitrise commented:
> ate the workflow to run the tests?

Tests are already run, I updated the workflow to deploy the results, so we can always check them.

richard-bogdan-bitrise commented:
Updated code with the suggested changes, also added the fix for case sensitivity issue, please check the updates 🙂
